File: 315_SmallerAfterSelf.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Stack;

class SmallerAfterSelf {
  public static void main(String[] args) {
    SmallerAfterSelf obj = new SmallerAfterSelf();
    System.out.println(obj.countSmaller(new int[] {5,2,6,1}));
  }

  public List<Integer> countSmaller(int[] nums) {
    int n = nums.length;
    List<Integer> res = new ArrayList<>();
    Stack<Integer> stack = new Stack<>();
    BSTNode root=null;
    for (int i = n-1; i>=0; --i) {
      stack.push(BSTSearch(root, nums[i]));
      root = BSTinsert(root, nums[i]);
    }

    while (!stack.isEmpty())
      res.add(stack.pop());
    return res;
  }

  private int BSTSearch(BSTNode root, int val) {
    if (root==null) return 0;

    if (root.val==val) return root.smallcount;
    else if (root.val > val)
      return BSTSearch(root.left, val);
    else
      return 1 + root.smallcount + BSTSearch(root.right, val);
  }

  private BSTNode BSTinsert(BSTNode root, int val) {
    if (root==null) return new BSTNode(0, val);
    
    if (root.val>val) {
      root.left = BSTinsert(root.left, val);
      root.smallcount++;
      return root;
    }
    else {
      root.right = BSTinsert(root.right, val);
      return root;
    }
  }

  class BSTNode extends TreeNode {
    int smallcount;
    BSTNode left;
    BSTNode right;
    BSTNode(int sc, int val) {
      super(val);
      smallcount = sc;
    }
  }



  public List<Integer> countSmallerBS(int[] nums) {
    int n = nums.length;
    List<Integer> res = new ArrayList<>();
    Stack<Integer> stack = new Stack<>();
    for (int i = n-1; i>=0; --i) {
      Arrays.sort(nums, i+1, n);
      int loc = Arrays.binarySearch(nums, i+1, n, nums[i]);
      if (loc<0) loc = - loc - 1;
      do {
        loc--;
      } while (loc>=i && nums[i]<=nums[loc]);

      stack.push(Math.max(loc-i, 0));
    }

    while (!stack.isEmpty())
      res.add(stack.pop());
    return res;
  }




  public List<Integer> countSmallerST(int[] nums) {
    List<Integer> res = new ArrayList<>();
    SegmentTree st = new SegmentTree(nums);
    for (int i=0; i<nums.length; i++) {
      res.add(st.RMQ(0, st.n-1, i+1, st.n-1, 0));
    }
    return res;
  }

  class SegmentTree {
    int[] st, stmin, arr;
    int n;

    SegmentTree(int[] arr) {
      n = arr.length;
      this.arr = arr;
      int power = (int) Math.ceil(Math.log(n) / Math.log(2)) + 1;
      int maxsize = (1 << (power)) - 1;
      st = new int[maxsize];
      stmin = new int[maxsize];
      
      if (n>0) {
        constructUtil(arr, 0, n-1, 0);
      }
    }

    int getMid(int s, int e) {
      return s+(e-s)/2;
    }

    int constructUtil(int[] arr, int ss, int se, int si) {
      if (ss==se) {
        st[si] = arr[ss];
        return arr[ss];
      }

      int mid = getMid(ss, se);
      st[si] = Math.max(constructUtil(arr, ss, mid, si * 2 + 1),
                        constructUtil(arr, mid+1, se, si * 2 + 2));
      return st[si];
    }

    int constructUtilMin(int[] arr, int ss, int se, int si) {
      if (ss==se) {
        stmin[si] = arr[ss];
        return arr[ss];
      }

      int mid = getMid(ss, se);
      stmin[si] = Math.min(constructUtil(arr, ss, mid, si * 2 + 1),
                        constructUtil(arr, mid+1, se, si * 2 + 2));
      return stmin[si];
    }

    int RMQ(int ss, int se, int qs, int qe, int si) {
      if (qs<=ss && qe>=se)
        if (st[si]<arr[qs-1])
          return se - ss + 1;
        else if (stmin[si]>=arr[qs-1] || ss==se) return 0;
      
      if (se<qs || ss>qe) return 0;

      int mid = getMid(ss, se);
      return RMQ(ss, mid, qs, qe, si * 2 + 1) + RMQ(mid+1, se, qs, qe, si * 2 + 2);
    }
  }
}

File: 315_SmallerAfterSelf.cpp
#include "utils.hpp"

class SegmentTree {
  private:
    int n;
    vector<int> stmax, stmin;
    const vector<int>& arr;

    int getMid(int ss, int se) {
      return ss + (se-ss)/2;
    }

    int constructUtil(int ss, int se, int si, 
                      vector<int> SegmentTree::* st, 
                      const int& (*comp)(const int&, const int &)) {
      // ...
      if (ss==se) {
        (this->*st)[si] = arr[ss];
        return arr[ss];
      }

      int mid = getMid(ss, se);
      (this->*st)[si] = comp(
        constructUtil(ss, mid, si*2+1, st, comp),
        constructUtil(mid+1, se, si*2+2, st, comp)
      );

      return (this->*st)[si];
    }


    int queryUtil(int ss, int se, int qs, int qe, int si) {
      if (ss>=qs && se<=qe) {
        if (stmax[si] < arr[qs-1])
          return se-ss+1;
        else if (stmin[si] >= arr[qs-1] || ss==se)
          return 0;
      }
      else if (se<qs || ss>qe)
        return 0;

      int mid = getMid(ss, se);
      return queryUtil(ss, mid, qs, qe, si*2+1) + queryUtil(mid+1, se, qs, qe, si*2+2);
    }

  public:
    SegmentTree(const vector<int>& _arr): arr(_arr) {
      n = arr.size();

      int i=0;
      while ((1 << i) < n)
        ++i;

      int maxsize = (1 << (i+1)) - 1;
      stmax.resize(maxsize);
      stmin.resize(maxsize);

      if (n>0) {
        constructUtil(0, n-1, 0, &SegmentTree::stmax, max);
        constructUtil(0, n-1, 0, &SegmentTree::stmin, min);
      }
    }

    int query(int qs) {
      return queryUtil(0, n-1, qs, n-1, 0);
    }
};


struct BSTNode 
  : public TreeNode {
  // ...
  int small, count;
  BSTNode(int _v): small(0), count(1), TreeNode(_v) { }
};


class ST {
  private:
    int n;
    vector<int> st;

    int getMid(int ss, int se) {
      return ss + (se-ss)/2;
    }

    void updateUtil(int ss, int se, int ui, int si) {
      if (ui<ss || ui>se)
        return;

      ++st[si];
      if (ss!=se) {
        int mid = getMid(ss, se);
        updateUtil(ss, mid, ui, si*2+1);
        updateUtil(mid+1, se, ui, si*2+2);
      }
    }

    int queryUtil(int ss, int se, int qs, int qe, int si) {
      if (ss>=qs && se<=qe)
        return st[si];
      else if (ss>qe || se<qs)
        return 0;

      int mid = getMid(ss, se);
      return queryUtil(ss, mid, qs, qe, si*2+1) +
             queryUtil(mid+1, se, qs, qe, si*2+2);
    }
  
  public:
    ST(int _n): n(_n) {
      int i=0;
      while ((1<<i) < n)
        ++i;

      st.resize((1 << (i+1)) - 1);
    }

    void update(int ui) {
      updateUtil(0, n-1, ui, 0);
    }

    int query(int qe) {
      return queryUtil(0, n-1, 0, qe, 0);
    }
};


class Solution {
  private:
    BSTNode* insert(TreeNode* root, int val) {
      if (!root)
        return new BSTNode(val);

      BSTNode* node = static_cast<BSTNode*>(root);
      if (val < node->val) {
        node->left = insert(node->left, val);
        ++node->small;
      }
      else if (val > root->val) {
        node->right = insert(node->right, val);
      }
      else
        ++node->count;

      return node;
    }

    int query(TreeNode* root, int val) {
      if (!root)
        return 0;

      BSTNode* node = static_cast<BSTNode*>(root);
      if (val < node->val)
        return query(node->left, val);
      else if (val > node->val)
        return node->count + node->small + query(node->right, val);
      else
        return node->small;
    }

  public:
    vector<int> countSmaller(vector<int>& nums) {
      int n = nums.size();
      vector<int> res(n);
      BSTNode* root = nullptr;
      for (int i=n-2; i>=0; --i) {
        root = insert(root, nums[i+1]);
        res[i] = query(root, nums[i]);
      }

      return res;
    }

    vector<int> countSmallerST(vector<int>& nums) {
      int n = nums.size();
      vector<int> res(n);

      if (n<2) 
        return res;

      int num_min = nums[0];
      int num_max = nums[0];

      for (int i=1; i<n; ++i)
        if (nums[i]<num_min)
          num_min = nums[i];
        else if (nums[i]>num_max)
          num_max = nums[i];

      ST st(num_max-num_min+1);
      for (int i=n-2; i>=0; --i) {
        st.update(nums[i+1]-num_min);
        res[i] = st.query(nums[i]-num_min-1);
      }

      return res;
    }

    vector<int> countSmallerTLE(vector<int>& nums) {
      // TLE...
      // between O(nlogn) and O(n2)
      // try alternative implementation of segment tree
      // given -10e4 <= num <= 10e4
      int n = nums.size();
      vector<int> res;
      res.reserve(n);

      SegmentTree st(nums);
      for (int i=0; i<n; ++i)
        res.push_back(st.query(i+1));

      return res;
    }
};


int main() {
  {
    UFUNC(Solution::countSmaller);
  }
  {
    UFUNC(Solution::countSmallerST);
  }
  {
    UFUNC(Solution::countSmallerTLE);
  }
  return 0;
}